pkg/loop/internal/ccip_execution.go

```diff
@@ -360,18 +358,16 @@ func (e *execProviderServer) NewOnRampReader(ctx context.Context, req *ccippb.Ne
 		return nil, err
 	}
 	// wrap the reader in a grpc server and serve it
-	srv := ccipinternal.NewOnRampReaderServer(reader)
+	srv := ccipinternal.NewOnRampReaderGRPCServer(reader)
 	// the id is handle to the broker, we will need it on the other side to dial the resource
 	onRampID, onRampResource, err := e.ServeNew("OnRampReader", func(s *grpc.Server) {
 		ccippb.RegisterOnRampReaderServer(s, srv)
 	})
 	if err != nil {
 		return nil, err
 	}
-	// TODO BCF-3067 LEAKS!!!
-	// this dependency needs to be closed when the onramp reader is closed, which
-	// should happen when the calling reporting plugin is closed/goes out of scope
-	e.deps.Add(onRampResource)
+	// ensure the grpc server is closed when the onRamp is closed. See comment in NewPriceRegistryReader for more details
+	srv.AddDep(onRampResource)
```

Comment:
this is the meat of the change. same pattern is implemented for the all the other services already